### ISSUE TITLE
Why  ORA-01406 fetched column value was truncated?

### ISSUE THREAD

**Issue by bangusi** [vrogier/ocilib#215]
My environment
Windows 10
Visual Studio 2017
Database Oracle 12c
Character set:   WE8MSWIN1252
ociliba compiled in x64-bit, static build

My application internally stores string data in UTF8
When I query a table occasionally I get this error ORA-01406
Partial code:
```
while ( rset++ )
{
  std::string val = rset.Get<std::string> ( pos );
}
```
Why is truncation occurring? 
Further digging shows the error is not caused by one particular row  but a set of rows, that is, individual rows from the set can be fetched successfully but when selected together the error occurs.

**Comment by vrogier** [comment]
Hi,

Can you provide sample data and code reproducing the issue?

Thanks

Vincent

**Comment by bangusi** [comment]
It is hard to provide the data that will re-produce this.
Unfortunately I don't own the data and I wouldn't be allowed to share it.
I have tried to mock-up similar data but can't re-produce the error ( I ddon't have a WE8MSWIN1252 database, mine is UTF8)
Hopefully these details might help:

- Database Characterset: WE8MSWIN1252
- Data type varchar2(30)
- Contains lots of French accented characters.  â, ê, î, ô, û,
- Some single characters and symbols e.h. +, / ** etc.
- 40,000 rows in the table
- 20,000 distinct values in the offending column.

My code is nothing special.

```
ocilib::Statement stmt ( connection );
stmt.SetFetchSize ( 150 );
stmt.Execute ( "select description from codes_table"  );
ocilib::Resultset rset = stmt.GetResultset();
while ( rset++ )
{
	std::string val = rset.Get<std::string> ( pos );
}
```
I am thinking switching to `ocilibw.lib` instead of `ociliba.lib` might work around the issue but that would require huge changes to my code.
If I find more useful info I will post here.

**Comment by vrogier** [comment]
Ok, I will try to reproduce it given your inputs.

Just to be sure. 
On application side, you use std::string as character container and use utf8 as character encoding.
On DB side, you use windows default encoding WE8MSWIN1252 (CP1252).
What is your nls_lang settings?

The truncation error means that the buffer provided by ocilib to receive the data is too small. 
Which implies that a conversion is done somewhere by oracle code.

**Comment by vrogier** [comment]
How oracle know ms that your string content is encoded in utf8 (in order to encode/decode properly to/from cp1252) ?

**Comment by bangusi** [comment]
I compiled a simple program in OCI_CHARSET_WIDE mode
and queried the same table successfully.. no truncation
Any ideas on how I can pin down where the truncation is happening.
My preference is to continue using UTF8

**Comment by vrogier** [comment]
Hi,

OCILIB checks for NLS_LANG environment variable. 
If it is set to any UTF8 values, then it allocates string buffers 4 times (worse case) the column length in order to have buffer big enough to receive data from oracle client (unless the column size semantics is in bytes)
OCILIB uses the C library getenv() method to retrieve environment variables.
It appears that visual studio implementation of getenv() sees environment variables that exist prior the start of the process and also the ones set by the process using putenv() method.
BUT not the ones set by the process using SetEnvironmentVariable() Windows method !!! 
Thus, it does not see the variable you set at the beginning of your program.
But if you do instead

`  putenv("NLS_LANG=AMERICAN_AMERICA.AL32UTF8");`

then it will work and you won't have the error anymore as OCILIB will use bigger buffer.

I could document it (e.g. "use putenv instead of  SetEnvironmentVariable on windows if you have to set programmatically") or add some code that call GetEnvironmentVariable() on windows if getenv() returned NULL)

**Comment by vrogier** [comment]
HI,

I've committed a workaround for it (it is not an ocilib issue, rather a msvc issue) in the v4.6.3 branch.

Regards,

Vincent

**Comment by bangusi** [comment]
The workaround does indeed work. Thanks.

**Comment by vrogier** [comment]
Hi, regarding these particular strings, can you provide an example

Vincent

**Comment by bangusi** [comment]
On Windows at least , using OCI_CHARSET_WIDE seems to be necessary if you want to avoid trouble.
[sample.txt](https://github.com/vrogier/ocilib/files/3791582/sample.txt)

Try this:
create table  large_text( txt clob);
load file content of file into one row of the table
query the table
using this pair of occilib programs and compare the output
	**-- ANSI --**
```
	std::string sql = "select txt from zz_large_txt";
	ocilib::Statement stmt(db_connection);
	stmt.Execute(sql);
	ocilib::Resultset rs = stmt.GetResultset();
	while (rs++)
	{
		auto val = rs.Get<std::string>(1);
	}
```
	
	**--WIDE--**
	
```
	std::wstring sql = L"select txt from zz_large_txt";
	ocilib::Statement stmt(db_connection);
	stmt.Execute(sql);
	ocilib::Resultset rs = stmt.GetResultset();
	while (rs++)
	{
		auto val =  rs.Get<std::wstring>(1);
	}
```

**Comment by vrogier** [comment]
Hi,


If NLS_LANG is not set to UTF8, you get the clob content into the string in ansi mode.
If NLS_LANG is set to UTF8, then the clob to string conversion returns an empty string!

I found the issue. When UTF8 is used, the clob to string conversion is correctly done.
But at the end of the conversion function, because of the stupid bug, we set the char of the buffer to 0.

I will close this issue as original scope was about the environment variable set using Win32 function and will open another issue and commit a fix + unit for it today.

Regards,

Vincent
